metaseq/models/transformer.py

```diff
@@ -561,10 +562,35 @@ def forward_embedding(
         incremental_state: Optional[Dict[str, Dict[str, Optional[Tensor]]]] = None,
     ):
         # embed tokens and positions
-        positions = None
+        if self.self_attn_doc_sep != -1:
+            # create own positions when self_attn_doc_sep is set
```

Comment:
I dont understand why this differs from the normal case?

Reply:
We are essentially resetting positions based on document separator tokens...

For instance, if the doc separator is 2, and the tokens are 
143 345 2 5435 2
The default positions would be
0 1 2 3 4

But with document level attention, we would like to reset positions as well on sentence boundaries. So, the positions become
0 1 2 0 1

Reply:
Hm. But if we're going this far, why not just have properly separate batch examples? The padding overhead, if implemented with heuristics, doesn't remotely approach high enough to worry, no?

Reply:
When we pack multiple examples in a sequence, we can efficiently use the full sequence length (2048), thereby each batch can accommodate more examples in comparison to having separate batch examples?